safe_authenticator/src/apps.rs

```diff
@@ -0,0 +1,237 @@
+                .nonce()
+                .ok_or_else(|| AuthError::from("No nonce on access container's MDataInfo"))?;
+
+            for app in auth_cfg.values() {
+                let key = access_container_enc_key(&app.info.id, &app.keys.enc_key, nonce)?;
+
+                // Empty entry means it has been deleted
+                let entry = match entries.get(&key) {
+                    Some(entry) if !entry.content.is_empty() => Some(entry),
+                    _ => None,
+                };
+
+                if let Some(entry) = entry {
+                    let plaintext = symmetric_decrypt(&entry.content, &app.keys.enc_key)?;
+                    let app_access = deserialise::<AccessContainerEntry>(&plaintext)?;
+
+                    let mut containers = HashMap::new();
+
+                    for (container_name, (_, permission_set)) in app_access {
+                        unwrap!(containers.insert(container_name, permission_set));
+                    }
+
+                    let registered_app = RegisteredApp {
+                        app_info: app.info.clone(),
+                        containers,
+                    };
+
+                    apps.push(registered_app);
+                }
+            }
+            Ok(apps)
+        })
+        .into_box()
+}
+
+/// Returns a list of applications that have access to the specified Mutable Data.
+pub fn apps_accessing_mutable_data(
+    client: &AuthClient,
+    name: XorName,
+    type_tag: u64,
+) -> Box<AuthFuture<Vec<AppAccess>>> {
+    let c2 = client.clone();
+
+    client
+        .list_mdata_permissions(name, type_tag)
+        .map_err(AuthError::from)
+        .join(config::list_apps(&c2).map(|(_, apps)| {
+            apps.into_iter()
+                .map(|(_, app_info)| (app_info.keys.sign_pk, app_info.info))
+                .collect::<HashMap<_, _>>()
+        }))
+        .and_then(move |(permissions, apps)| {
+            // Map the list of keys retrieved from MD to a list of registered apps (even if
+            // they're in the Revoked state) and create a new `AppAccess` struct object
+            let mut app_access_vec: Vec<AppAccess> = Vec::new();
+            for (user, perm_set) in permissions {
+                if let Key(public_key) = user {
+                    let app_access = match apps.get(&public_key) {
+                        Some(app_info) => AppAccess {
+                            sign_key: public_key,
+                            permissions: perm_set,
+                            name: Some(app_info.name.clone()),
+                            app_id: Some(app_info.id.clone()),
+                        },
+                        None => {
+                            // If an app is listed in the MD permissions list, but is not
+                            // listed in the registered apps list in Authenticator, then set
+                            // the app_id and app_name fields to ptr::null(), but provide
```

Comment:
`ptr::null()` should be `None` here I suppose:
```suggestion
                            // the app_id and app_name fields to None, but provide
```